### PR TITLE
Collect crc errors stats

### PR DESCRIPTION
**Pull request by netmaniac**
Count all messages received from SDS (both command replies and processed data packets) and checksum errors.

### CODE REVIEW

File: src/Sds011.cpp [hunk 1]
```diff
@@ -138,9 +138,16 @@ bool Sds011::crc_ok() {
     for (int i = 2; i < 8; ++i) {
         crc += _buf[i];
     }
+    if (crc != _buf[8]) checksum_errors++;
```
Comment (dok-net) [inline src/Sds011.cpp]: I prefer pre-increment unless the effect of post-increment is explicitly required.

File: src/Sds011.cpp [hunk 2]
```diff
@@ -138,9 +138,16 @@ bool Sds011::crc_ok() {
...
     return crc == _buf[8];
 }
 
+bool Sds011::crc_stats(uint32_t& err, uint32_t& total){
```
Comment (dok-net) [inline src/Sds011.cpp]: Blank between ) and {

File: src/Sds011.cpp [hunk 3]
```diff
@@ -138,9 +138,16 @@ bool Sds011::crc_ok() {
...
+    err = checksum_errors;
+    total = processed_responses;
+    return true;
+};
+
```
Comment (dok-net) [inline src/Sds011.cpp:150]: References to ints, I don't like that, you could either return a struct, or just provide two accessors.
Also, I would like to be able to reset the counters.

**Review by dok-net — COMMENTED** [review]
Very good, I was obviously thinking about this myself as as result of your issue report, so I am absolutely willing to merge this when you've ironed out a few little things. As long as the reported values are exactly what you need, I am fine with what you're counting, haven't given it any deep sanity check myself, probably at the end, but until then I trust you know what you need.

File: src/Sds011.cpp (hunk 3 above)
Comment (netmaniac) [reply src/Sds011.cpp:150]: I was copying style from other functions like `query_data_auto`

**Comment by dok-net** [comment]
@netmaniac You are just one step away ... I'll gladly merge ...

**Comment by dok-net** [comment]
I am getting 0 errors running `measure.ino` extended to display the new CRC reporting:
```
Serial RX CRC errors / total frames: 0 / 2140
```
Edit: see below, turns out that I believe this metric is too optimistic.

**Comment by dok-net** [comment]
Wait a minute... just had a thought off the top of my head... this PR counts CRC fails in one place, and "processed responses" in another. What about any frames that don't make it to the crc check, those are fails, too, no?

**Comment by dok-net** [comment]
I am very sorry, I have to stand corrected: your PR seems very flawed to me, you have the error to total frame received logic wrong and the results would be totally overoptimistic. If you can prove me wrong, I'm all for it, but from my more in depth review it certainly transpires that all dropped or partial frames, that don't even make it to CRC checking, are disregarded, which causes a total bias in favor of good RX quality when there is none.
You need to properly count all receives, and report how much of that is "bad", to give a meaningful indication of anything.
Are you with me on this?

**Comment by netmaniac** [comment]
I'm away for couple of days. I'll get back to You in few days.

**Comment by netmaniac** [comment]
What I try to count is number of dropped frames due to bad checksum. Checksum covers only what is inside correct frame. The frame is defined by the start and stop sequence. If there is error in start/stop sequence then it should not be covered as checksum error.

**Comment by dok-net** [comment]
It just seems to give a poor measurement of communications quality overall, when done this way. You are being consistent to the wording, but I would rather that any kind of bad frame is counted as well, like those due to missing bytes, which time out.

**Comment by dok-net** [comment]
@netmaniac What do you think, are you considering adding more comprehensive error counting? Because, if I am not mistaken, most if not all byte losses cause the frame discard logic to kick in, which in turn fails to add those errors to the statistics, so the report egregiously errs on the positive side.

**Comment by netmaniac** [comment]
I'm sorry but I'm currently involved in different topics and don't see possibility to continue with this.

**Comment by dok-net** [comment]
No worries, don't delete your work, the code base doesn't change that rapidly, you can get back to this at a later time.

**Comment by dok-net** [comment]
This PR is stale.

**Pull request closed without merging**